Add --wait option to start-build

mfojtik opened the pull request:
This will add support for `oc start-build foo --wait`. In that case, the `start-build` command will start the build and waits till the build is complete. If this is true, the command exits with 0, if the build failed or the Phase is not Complete, this command will exit with 1.

Note that you can still specify the `--follow` option which will stream the logs. Together with `--wait` it will get the behavior I described above.

Usage:
```
# This will stream the logs to console, but exit always with 0
$ oc start-build sample-app --follow

# This will wait for the build to complete and stream the logs to console. 
# It exists with 1 if the build failed
$ oc start-build sample-app --follow --wait

# This will wait for the build to complete and exit 1 when the build failed
$ oc start-build sample-app --wait
```

Sample error output:
```
$ oc start-build sample-app --wait ; echo $?
sample-app-6
The build dev/sample-app-6 status is "Failed"
1
```

jhadvig commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -122,21 +127,49 @@ func RunStartBuild(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args
 	}
+
+	wg.Wait()
+	// When waiting for build to complete the build fail, exit with 1 so the
```
... for the build to complete and the build fails, ....

jhadvig commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -60,6 +63,7 @@ func NewCmdStartBuild(fullName string, f *clientcmd.Factory, out io.Writer) *cob
 	}
 	cmd.Flags().String("from-build", "", "Specify the name of a build which should be re-run")
 	cmd.Flags().Bool("follow", false, "Start a build and watch its logs until it completes or fails")
+	cmd.Flags().Bool("wait-for-complete", false, "Wait for a build to complete (will return error if build fails)")
```
just an cosmetic suggestion :) , nuke the parentheses and rather use the semicolon as the example bellow.

mfojtik commented:
@bparees @fabianofranz @smarterclayton PTAL

I have tested this manually and all three cases works perfectly.

mfojtik commented:
Fixes: https://github.com/openshift/origin/issues/4557

smarterclayton commented:
--wait would be a lot more succinct.  I don't think we'd have a variant "wait for failure".

mfojtik commented:
@smarterclayton +1 --wait is shorter thus better :-)

mfojtik commented:
@smarterclayton fixed.

bparees commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -40,7 +43,12 @@ This command starts a build for the provided BuildConfig or re-runs an existing
+  $ %[1]s start-build 3bd2ug53b --follow
+
+  // Starts build from BuildConfig matching the name "3bd2ug53b" and wait until
+	// the build completes. It exits with non-zero return code if the build
```
with a non-zero return code

bparees commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -61,6 +69,7 @@ func NewCmdStartBuild(fullName string, f *clientcmd.Factory, out io.Writer) *cob
 	cmd.Flags().String("from-build", "", "Specify the name of a build which should be re-run")
 	cmd.Flags().String("commit", "", "Specify the commit hash the build should be run from")
 	cmd.Flags().Bool("follow", false, "Start a build and watch its logs until it completes or fails")
+	cmd.Flags().Bool("wait", false, "Wait for a build to complete and exit with non-zero return node if the build fails")
```
with A non-zero return Code

bparees commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -337,3 +375,50 @@ func gitRefInfo(repo git.Repository, dir, ref string) (buildapi.GitRefInfo, erro
+		return b.Status.Phase == buildapi.BuildPhaseComplete
+	}
+	isFailed := func(b *buildapi.Build) bool {
+		return b.Status.Phase == buildapi.BuildPhaseFailed || b.Status.Phase == buildapi.BuildPhaseError
```
cancelled?

mfojtik commented:
@bparees thanks! all comments fixed.

bparees commented:
test in test-cmd?

bparees commented:
(or extended test)

mfojtik commented:
@bparees need to refactor test-end-to-end.sh to use this... can be follow-up or part of this PR :-)

mfojtik commented:
@bparees ok, so the start-build invocation is commented out in test-end-to-end and I don't really want to write an extended test to execute build and wait for it ;-) Any suggestions? I guess the test-end-to-end.sh use BuildConfig wit ConfigChangeTrigger so there is no need to start-build.

bparees commented:
you could add a start-build to e2e after the first build finishes.  that should be fast so it's a good place for it.

but you also need a failing case.

mfojtik commented:
@bparees maybe I can specify non-existing --commit as well for a failure case.

bparees commented:
@mfojtik or you could cancel the build immediately after starting it.

you should probably do all of these things. :)

mfojtik commented:
@bparees https://gist.github.com/mfojtik/62d4e15fa7125f8fda0a ?

mfojtik commented:
@bparees don't want to cancel, want to see it dying terribly :-)

mfojtik commented:
[test]

bparees commented:
@mfojtik you should still also do the cancel test, so you're covering all the code paths.

mfojtik commented:
@bparees there is no 'easy' way to exercise `cancel-build` on something that was started by `start-build --wait` in the same script:

1) I will have to background the --wait and redirect the output to a file to get the build name
2) I will have to call cancel-build with that name AND also wait for the start-build (get the return code)

This will be by definition very flaky... I think we should do the cancel-build testing in extended test (and keep the --commit and --wait in test-cmd.sh as they are easy to test). I can do that as a follow up.

bparees commented:
>  I think we should do the cancel-build testing in extended test (and keep the --commit and --wait in test-cmd.sh as they are easy to test). I can do that as a follow up.

works for me.

mfojtik commented:
@bparees and we are green in Jenkins!

bparees commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -40,7 +43,12 @@ This command starts a build for the provided BuildConfig or re-runs an existing
-  $ %[1]s start-build 3bd2ug53b --follow`
+  $ %[1]s start-build 3bd2ug53b --follow
+
+  // Starts build from BuildConfig matching the name "3bd2ug53b" and wait until
```
how did we pick such a horrible example build name?

kargakis replied:
upstream examples c/p most probably

mfojtik replied:
we should start a project named '3bd2ug53b' :-)

bparees commented on hack/test-end-to-end.sh:
```diff
@@ -196,6 +196,11 @@ wait_for_build_start "test"
+echo "[INFO] Starting build from ${STI_CONFIG_FILE} with non-existing commit..."
+set +e
+oc start-build test --commit=fffffff --wait && echo "The build was supposed to fail, but it succeeded." && exit 1
+set -e
```
that's the negative test, where's the positive test?

mfojtik replied:
@bparees not sure we want to add one as it will make the end-to-end longer (+~20s)... 
I think the test where we actually wait for build to fail is enough... I will add extended test for cancel anyway, so I can add one extra to verify the wait return 0 when build succeed.

bparees replied:
as long as you promise to add it in an extended test.

bparees commented:
- missing a positive test (rc=0) for --wait
- needs squash
- minor nit about the usage (which i realize you didn't originate but you can fix :) )

kargakis commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -337,3 +375,52 @@ func gitRefInfo(repo git.Repository, dir, ref string) (buildapi.GitRefInfo, erro
 }
+
+// WaitForBuildComplete waits for a build identified by the name to complete
+func WaitForBuildComplete(c osclient.BuildInterface, name string) error {
```
I see duplicate code with the buildlog registry

Can't you factor out common parts to util functions? Follow-up is fine

mfojtik replied:
Yes, there is another occurrence in extended tests... we should move this to some common helper, will do as follow up.

kargakis replied:
Actually now that I noticed, `start-build` calls `BuildLogs().Get` so essentially this code runs twice both in the client and the server. I would propose tweaks to the server code instead of duplicate code running in parallel:)

kargakis replied:
https://github.com/openshift/origin/blob/90750d8a2066c6c7469bef0419649d2c4d6555cc/pkg/build/registry/buildlog/rest.go#L113

mfojtik replied:
@kargakis you can use --wait or --follow or both... In case you use both, then yeah, the code runs twice on server side and in client. However, if you just use --wait the code runs just in CLI (and the logs are not streamed out). So I think this PR is correct.

kargakis replied:
Now I see. Then in case of a `start-build foo --follow --wait` the cli wait code shouldn't run since wait already happens in --follow.

kargakis replied:
Actually why someone would call `start-build foo --follow --wait`. Should these work together?

kargakis replied:
> # It exists with 1 if the build failed

So this is just for failed builds

kargakis commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -337,3 +375,52 @@ func gitRefInfo(repo git.Repository, dir, ref string) (buildapi.GitRefInfo, erro
+	isOK := func(b *buildapi.Build) bool {
+		return b.Status.Phase == buildapi.BuildPhaseComplete
+	}
+	isFailed := func(b *buildapi.Build) bool {
```
This is incosistent with what happens on the server where a failed build isn't considered an error: https://github.com/openshift/origin/blob/90750d8a2066c6c7469bef0419649d2c4d6555cc/pkg/build/registry/buildlog/rest.go#L136

mfojtik replied:
this is intentional in CLI. If the build failed, you want to make `oc start-build --wait` return 1 instead of 0, so the script you are running this can make decision based on that status.

In Origin Failed state means the build is done, in start-build --wait it means the command failed.

mfojtik replied:
@kargakis also the server function is meant to give you logs when the build is running, complete or failed whereas the function I have cares just about Complete/Failed/Cancelled... so I guess this is a different function.... But I agree that we should have something generic we can re-use.... WaitForBuild(BuildInterface, okCondition, failedCondition).... I think we already have this in extended tests, need to move that to Origin.

kargakis commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -132,21 +142,49 @@ func RunStartBuild(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args
+	)
+
+	// Wait for the build to complete
+	if waitForComplete {
```
```go
if waitForComplete && !follow {
```

bparees replied:
you need to do this even when specifying follow so that the exit code is reported.

kargakis commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -33,14 +36,19 @@ This command starts a build for the provided BuildConfig or re-runs an existing
+  $ %[1]s start-build sample-build --follow
+
+  // Starts build from BuildConfig matching the name "sample-build" and wait until
+	// the build completes. It exits with a non-zero return code if the build
```
Why these comments have different indentation?

bparees commented on docs/generated/oc_by_example_content.adoc:
```diff
@@ -792,6 +792,11 @@ Starts a new build
+  // Starts build from BuildConfig matching the name "3bd2ug53b" and wait until
+	// the build completes. It exits with a non-zero return code if the build
+	// fails. 
+  $ openshift cli start-build 3bd2ug53b --wait
```
need to regen docs?

mfojtik commented:
OK guys, so I updated the docs, fixed indentation of the comment and going to merge this.

As a follow up, couple things will happen:

* I will add extended test to exercise various ways you can start/cancel the build.
* I will remove duplication of WaitForBuild() in server, extended test and here.

mfojtik commented:
[merge]

kargakis commented:
Fixed https://github.com/openshift/origin/issues/4557
